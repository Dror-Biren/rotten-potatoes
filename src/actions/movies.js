//import uuid from 'uuid';
import database from '../firebase/firebase';
import { uploadPoster } from "../firebase/posters";


// ADD_MOVIE
export const addMovie = (movie) => ({
   type: 'ADD_MOVIE',
   movie
});

function addMovieToDatabase(movie, dispatch) {
   return database.ref(`movies`).push(movie)
      .then((ref) =>
         dispatch(addMovie({
            id: ref.key,
            ...movie
         }))
      );
}

export const startAddMovie = (movieData = {}) => {
   return (dispatch, getState) => {
      const {
         title = '',
         description = '',
         rating = 0,
         ratingsAmount = 0,
         releaseDate = 0,
         genres = [],
         poster,
         trailerUrl = ''
      } = movieData;

      const creator = {
         uid: getState().auth.uid,
         time: new Date().valueOf()
      }

      const movie = { title, description, rating, ratingsAmount, releaseDate, genres, trailerUrl, creator, raters: {} }

      function afterUploadPoster(newUrl) {
         movie.posterUrl = newUrl;
         return addMovieToDatabase(movie, dispatch);
      }

      if (poster)
         return uploadPoster(poster)
            .then(afterUploadPoster)

      return addMovieToDatabase(movie, dispatch);
   }
}




function editMovieInDatabase(updates, id, dispatch, newRater) {
   let firebaseUpdates = updates;
   if (newRater) {
      const newRaterPath = `raters/${newRater.key}`;
      firebaseUpdates[newRaterPath] = newRater.data;
   }

   return database.ref(`movies/${id}`).update(firebaseUpdates).then(() => {
      dispatch(editMovie(id, updates, newRater));
   });
}

// EDIT_MOVIE
export const editMovie = (id, updates, newRater) => ({
   type: 'EDIT_MOVIE',
   id,
   updates,
   newRater
});

export const startEditMovie = (id, updates) => {
   return (dispatch, getState) => {
      const { poster, newRatingVote } = updates;
      delete updates.poster;
      delete updates.newRatingVote;

      const newRater = newRatingVote && {
         key: getState().auth.uid,
         data: {
            rating: newRatingVote,
            time: new Date().valueOf()
         }
      }

      if (poster)
         return uploadPoster(poster)
            .then(afterUploadPoster)

      return editMovieInDatabase(updates, id, dispatch, newRater);

      function afterUploadPoster(newUrl) {
         updates.posterUrl = newUrl;
         return editMovieInDatabase(updates, id, dispatch, newRater);
      }
   };
};



// UPDATE_MOVIE_RATING
export const updateMovieRating = (movie, newRatingVote) => {
   let { id, rating, ratingsAmount } = movie;
   const newRatingSum = ratingsAmount * rating + newRatingVote;
   ratingsAmount++;
   rating = newRatingSum / ratingsAmount;
   return startEditMovie(id, { rating, ratingsAmount, newRatingVote })
}



// REMOVE_MOVIE
export const removeMovie = ({ id } = {}) => ({
   type: 'REMOVE_MOVIE',
   id
});

export const startRemoveMovie = ({ id } = {}) => {
   return (dispatch, getState) => {
      return database.ref(`movies/${id}`).remove().then(() => {
         dispatch(removeMovie({ id }));
      });
   };
};





// SET_MOVIES
export const setMovies = (movies) => ({
   type: 'SET_MOVIES',
   movies
});

export const startSetMovies = () => {
   return (dispatch, getState) => {
      return database.ref(`movies`).once('value').then((snapshot) => {
         const movies = [];

         snapshot.forEach((childSnapshot) => {
            movies.push({
               id: childSnapshot.key,
               ...childSnapshot.val()
            });
         });

         if (!movies.genres)
            movies.genres = [];

         dispatch(setMovies(movies));
      });
   };
};
