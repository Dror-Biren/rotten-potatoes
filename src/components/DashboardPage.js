import React from 'react';

import MoviesList from './MoviesList';
import MovieListFilters from './MovieListFilters';
import MoviesSummary from './MoviesSummary';

const DashboardPage = () => (
  <div>
    <MoviesSummary />
    <MovieListFilters />
    <MoviesList />
  </div>
);

export default DashboardPage;
