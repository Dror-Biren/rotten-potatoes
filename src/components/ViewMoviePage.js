import React from 'react';
import { connect } from 'react-redux';
import moment from 'moment';

import UserViewMovie from './UserViewMovie';
import AdminEditMovie from './AdminEditMovie';

export class ViewMoviePage extends React.Component {
   returnToMainPage = () => {
      this.props.history.push('/');
   };

   render() {
      const { movie } = this.props;
      const releaseYear = moment(movie.releaseDate).year();

      const adminContent = (
         <AdminEditMovie
            movie={movie}
            returnToMainPage={this.returnToMainPage}
         />
      );

      const userContent = (
         <UserViewMovie
            movie={movie}
            returnToMainPage={this.returnToMainPage}
         />
      );


      return (
         <div>
            <div className="page-header">
               <div className="content-container">
                  <h1 className="page-header__title">
                     {movie.title} &nbsp;&nbsp; ({releaseYear})
                  </h1>
               </div>
               <img
                  src="/images/goBack5.png"
                  className="goBack"
                  onClick={this.returnToMainPage}>
               </img>
            </div>
            <div className="content-container">
               {this.props.isAdmin ? adminContent : userContent}
            </div>
         </div>
      );
   }
};


const mapStateToProps = (state, props) => {
   const matchId = ({ id }) => id === props.match.params.id;
   return {
      movie: state.movies.find(matchId),
      isAdmin: state.auth.isAdmin
   }
};

export default connect(mapStateToProps)(ViewMoviePage);
