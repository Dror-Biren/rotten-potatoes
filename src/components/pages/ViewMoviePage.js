import React from 'react';
import { connect } from 'react-redux';
import moment from 'moment';

import UserViewMovie from './UserViewMovie';
import AdminEditMovie from './AdminEditMovie';

export class ViewMoviePage extends React.Component {
   returnToMainPage = () => {
      this.props.history.push('/dashboard');
   };

   render() {
      const { movie } = this.props;
      if (!movie) {
         this.props.history.push('/not-found-page');
         return <div></div>;
      }

      const releaseYear = moment(movie.releaseDate).year();

      const adminContent = (
         <AdminEditMovie
            movie={movie}
            returnToMainPage={this.returnToMainPage}
         />
      );

      const userContent = (
         <UserViewMovie
            movie={movie}
            returnToMainPage={this.returnToMainPage}
         />
      );


      return (
         <div>
            <div className="sub-header">
                  <h2 className="sub-header__title">
                     <span className="span-bold">
                        {movie.title}
                     </span> 
                     &nbsp; ({releaseYear})
                  </h2>
               <img
                  src="/images/goBack5.png"
                  className="button--go-back"
                  onClick={this.returnToMainPage}>
               </img>
            </div>
            <div className="page-margin page-content-padding moviePage-content">
               {this.props.isAdmin ? adminContent : userContent}
            </div>
         </div>
      );
   }
};


const mapStateToProps = (state, props) => {
   const matchId = ({ id }) => id === props.match.params.id;
   return {
      movie: state.movies.find(matchId),
      isAdmin: state.user.isAdmin
   }
};

export default connect(mapStateToProps)(ViewMoviePage);
