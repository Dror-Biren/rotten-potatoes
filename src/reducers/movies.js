// Expenses Reducer

const moviesReducerDefaultState = [];

export default (state = moviesReducerDefaultState, action) => {
   switch (action.type) {
      case 'ADD_MOVIE':
         return [
            ...state,
            action.movie
         ];

      case 'REMOVE_MOVIE':
         const isDiffrentMovie = ({ id }) => id !== action.id;
         return state.filter(isDiffrentMovie);

      case 'EDIT_MOVIE':
         const getEditedMovie = (movie) => {
            if (action.newRater) {
               const { key, data } = action.newRater;
               if (!movie.raters)
                  movie.raters = {};
               movie.raters[key] = data;
            }
            return {
               ...movie,
               ...action.updates
            };
         }

         return state.map((movie) => {
            if (movie.id === action.id)
               return getEditedMovie(movie);
            return movie;
         });     

      case 'SET_MOVIES':
         return action.movies;

      default:
         return state;
   }
};
