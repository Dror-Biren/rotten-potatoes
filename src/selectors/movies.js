import moment from 'moment'

// Get visible movies
export default (movies, { text, sortBy, startYear, endYear }) => {
  //console.log({startYear, endYear});

  function isMovieMatch(movie) {
    const releaseYear = moment(movie.releaseDate).year();

    const startYearMatch = startYear ? startYear <= releaseYear : true;
    const endYearMatch = endYear ? endYear >= releaseYear : true;
    const textMatch = movie.title.toLowerCase().includes(text.toLowerCase());

    return startYearMatch && endYearMatch && textMatch;
  }

  function sortMovies(movie1, movie2) {
    switch(sortBy) {
      case 'viewes':
        return movie2.ratingsAmount - movie1.ratingsAmount;
      case 'date':
        return movie2.releaseDate - movie1.releaseDate;
      case 'rating':
        return movie2.rating - movie1.rating;
    }
  }

  return movies.filter(isMovieMatch).sort(sortMovies);
};
