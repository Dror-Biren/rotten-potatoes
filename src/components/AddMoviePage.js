import React from 'react';
import { connect } from 'react-redux';
import MovieForm from './MovieForm';
import { startAddMovie } from '../actions/movies';

export class AddMoviePage extends React.Component {
   returnToMainPage = () => {
      this.props.history.push('/');
   };

   onSubmit = (movie) => {
      this.props.startAddMovie(movie);
      this.props.history.push('/');
   };

   render() {
      return (
         <div>
            <div className="sub-header">
               
                  <h2 className="sub-header__title">
                     Add a new movie to the site
                  </h2>
               
               <img
                  src="/images/goBack5.png"
                  className="button--go-back"
                  onClick={this.returnToMainPage}>
               </img>
            </div>
            <div className="page-margin">
               <MovieForm
                  onSubmit={this.onSubmit}
               />
            </div>
         </div>
      );
   }
}

const mapDispatchToProps = (dispatch) => ({
   startAddMovie: (movie) => dispatch(startAddMovie(movie))
});

export default connect(undefined, mapDispatchToProps)(AddMoviePage);
