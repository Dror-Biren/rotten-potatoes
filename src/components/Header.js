import React from 'react';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';

import { startLogin } from '../actions/auth';
import { startLogout } from '../actions/auth';
import { history } from '../routers/AppRouter';

export const Header = ({ isAuthenticated, startLogin, startLogout }) => {
   const logOutButton = (
      <button className="button" onClick={startLogout}>
         Log-out
      </button>
   );

   const logInButton = (
      <button className="button" onClick={startLogin}>
         Login with Google
      </button>
   );

   return (
      <header className="header">
         <img
            className="header__background"
            src="/images/potatoes.webp"
         />
         <div className="content-container">
            <div className="header__content">

               <Link className="header__title" to="/dashboard">
                  <h1>
                     Rotten Potatoes
                  </h1>
               </Link>

               { isAuthenticated ? logOutButton : logInButton }
            </div>
         </div>
      </header>
   );
}

const mapStateToProps = (state) => ({
   isAuthenticated: !!state.auth.uid
 });

const mapDispatchToProps = (dispatch) => ({
   startLogin: () => dispatch(startLogin()),
   startLogout: () => {
      dispatch(startLogout()).then();
      const returnToLoginPage = () => history.push('/');
      setTimeout(returnToLoginPage, 700);
   }
});

export default connect(mapStateToProps, mapDispatchToProps)(Header);
