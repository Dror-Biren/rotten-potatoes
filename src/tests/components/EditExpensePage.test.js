import React from 'react';
import { shallow } from 'enzyme';
import movies from '../fixtures/movies';
import { ViewMoviePage } from '../../components/ViewMoviePage';

let startEditMovie, startRemoveMovie, history, wrapper;

beforeEach(() => {
  startEditMovie = jest.fn();
  startRemoveMovie = jest.fn();
  history = { push: jest.fn() };
  wrapper = shallow(
    <ViewMoviePage
      startEditMovie={startEditMovie}
      startRemoveMovie={startRemoveMovie}
      history={history}
      movie={movies[2]}
    />
  );
});

test('should render ViewMoviePage', () => {
  expect(wrapper).toMatchSnapshot();
});

test('should handle startEditMovie', () => {
  wrapper.find('MovieForm').prop('onSubmit')(movies[2]);
  expect(history.push).toHaveBeenLastCalledWith('/');
  expect(startEditMovie).toHaveBeenLastCalledWith(movies[2].id, movies[2]);
});

test('should handle startRemoveMovie', () => {
  wrapper.find('button').simulate('click');
  expect(history.push).toHaveBeenLastCalledWith('/');
  expect(startRemoveMovie).toHaveBeenLastCalledWith({
    id: movies[2].id
  });
});
