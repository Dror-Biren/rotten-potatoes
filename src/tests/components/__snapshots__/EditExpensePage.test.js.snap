// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ViewMoviePage 1`] = `
<div>
  <div
    className="page-header"
  >
    <div
      className="content-container"
    >
      <h1
        className="page-header__title"
      >
        Edit Movie
      </h1>
    </div>
  </div>
  <div
    className="content-container"
  >
    <MovieForm
      movie={
        Object {
          "rating": 4500,
          "releaseDate": 345600000,
          "title": "Credit Card",
          "id": "3",
          "description": "",
        }
      }
      onSubmit={[Function]}
    />
    <button
      className="button button--secondary"
      onClick={[Function]}
    >
      Remove Movie
    </button>
  </div>
</div>
`;
